x/protocolpool/keeper/msg_server_test.go

```diff
@@ -337,3 +336,358 @@ func (suite *KeeperTestSuite) TestMsgClaimBudget() {
+				Expiry:     &expiry,
+			},
+			expErr:    true,
+			expErrMsg: "invalid authority",
+		},
+		"zero percentage": {
+			input: &types.MsgCreateContinuousFund{
+				Authority:  suite.poolKeeper.GetAuthority(),
+				Recipient:  recipientAddr.String(),
+				Percentage: math.LegacyNewDec(0),
+				Expiry:     &expiry,
+			},
+			expErr:    true,
+			expErrMsg: "percentage cannot be zero or empty",
+		},
+		"negative percentage": {
+			input: &types.MsgCreateContinuousFund{
+				Authority:  suite.poolKeeper.GetAuthority(),
+				Recipient:  recipientAddr.String(),
+				Percentage: negativePercentage,
+				Expiry:     &expiry,
+			},
+			expErr:    true,
+			expErrMsg: "percentage cannot be negative",
+		},
+		"invalid percentage": {
+			input: &types.MsgCreateContinuousFund{
+				Authority:  suite.poolKeeper.GetAuthority(),
+				Recipient:  recipientAddr.String(),
+				Percentage: math.LegacyNewDec(1),
+				Expiry:     &expiry,
+			},
+			expErr:    true,
+			expErrMsg: "percentage cannot be greater than or equal to one",
+		},
+		"invalid expiry": {
+			input: &types.MsgCreateContinuousFund{
+				Authority:  suite.poolKeeper.GetAuthority(),
+				Recipient:  recipientAddr.String(),
+				Percentage: percentage,
+				Expiry:     &invalidExpirty,
+			},
+			expErr:    true,
+			expErrMsg: "expiry time cannot be less than the current block time",
+		},
+		"all good": {
+			input: &types.MsgCreateContinuousFund{
+				Authority:  suite.poolKeeper.GetAuthority(),
+				Recipient:  recipientAddr.String(),
+				Percentage: percentage,
+				Expiry:     &expiry,
+			},
+			expErr: false,
+		},
+	}
+
+	for name, tc := range testCases {
+		suite.Run(name, func() {
+			suite.SetupTest()
+
+			_, err := suite.msgServer.CreateContinuousFund(suite.ctx, tc.input)
+			if tc.expErr {
+				suite.Require().Error(err)
+				suite.Require().Contains(err.Error(), tc.expErrMsg)
+			} else {
+				suite.Require().NoError(err)
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestCancelContinuousFund() {
+	testCases := map[string]struct {
+		preRun        func()
+		recipientAddr sdk.AccAddress
+		expErr        bool
+		expErrMsg     string
+		postRun       func()
+	}{
+		"empty recipient": {
+			preRun: func() {
+				percentage, err := math.LegacyNewDecFromStr("0.2")
+				suite.Require().NoError(err)
+				oneMonthInSeconds := int64(30 * 24 * 60 * 60) // Approximate number of seconds in 1 month
+				expiry := suite.ctx.BlockTime().Add(time.Duration(oneMonthInSeconds) * time.Second)
+				cf := types.ContinuousFund{
+					Recipient:  "",
+					Percentage: percentage,
+					Expiry:     &expiry,
+				}
+				err = suite.poolKeeper.ContinuousFund.Set(suite.ctx, recipientAddr, cf)
+				suite.Require().NoError(err)
+			},
+			expErr:    true,
+			expErrMsg: "empty address string is not allowed",
+		},
+		"no recipient found": {
+			recipientAddr: recipientAddr,
+			expErr:        true,
+			expErrMsg:     "no recipient found to cancel continuous fund",
+		},
+		"all good": {
```

Comment:
what happens if the user has funds that were not claimed? should write a test case for this